### PR TITLE
Rebase onto Kubernetes 6241a21

### PR DESCRIPTION
**Pull request by smarterclayton**
Includes remote command execution and Docker caching

### CODE REVIEW

**Comment by smarterclayton** [comment]
[test]

**Comment by smarterclayton** [comment]
@pweil- @deads2k review please.

**Comment by smarterclayton** [comment]
Needs #1135 fixed upstream (although patched here).

**Comment by smarterclayton** [comment]
Might fix #1115

**Comment by smarterclayton** [comment]
I cherry-picked Vish's change in as well.

File: Godeps/_workspace/src/github.com/GoogleCloudPlatform/kubernetes/pkg/api/rest/create_test.go [hunk 1]
```diff
@@ -25,17 +25,17 @@ import (
 
 func TestCheckGeneratedNameError(t *testing.T) {
 	expect := errors.NewNotFound("foo", "bar")
-	if err := CheckGeneratedNameError(Pods, expect, &api.Pod{}); err != expect {
+	if err := CheckGeneratedNameError(Services, expect, &api.Pod{}); err != expect {
```
Comment (pweil-) [inline Godeps/_workspace/src/github.com/GoogleCloudPlatform/kubernetes/pkg/api/rest/create_test.go]: Is it correct to be using the services strategy when validating pod name generation?  In controllers/types.go they're  technically the same strategy, just struck me as strange when reading it.

Comment (pweil-) [inline Godeps/_workspace/src/github.com/GoogleCloudPlatform/kubernetes/pkg/api/rest/create_test.go]: Disregard, the next file removes the Pods strategy....

File: Godeps/_workspace/src/github.com/GoogleCloudPlatform/kubernetes/pkg/api/types.go [hunk 2]
```diff
@@ -734,12 +745,25 @@ type Service struct {
...
+// Name: "mysql", Endpoints: [{"ip": "10.10.1.1", "port": 1909}, {"ip": "10.10.2.2", "port": 8834}]
 type Endpoints struct {
 	TypeMeta   `json:",inline"`
 	ObjectMeta `json:"metadata,omitempty"`
 
-	Endpoints []string `json:"endpoints,omitempty"`
+	// Optional: The IP protocol for these endpoints. Supports "TCP" and
+	// "UDP".  Defaults to "TCP".
+	Protocol  Protocol   `json:"protocol,omitempty"`
+	Endpoints []Endpoint `json:"endpoints,omitempty"`
```
Comment (pweil-) [inline Godeps/_workspace/src/github.com/GoogleCloudPlatform/kubernetes/pkg/api/types.go]: I believe this is going to require some changes in the router (at least).  I'll test after merge and submit a PR.

**Comment by pweil-** [comment]
My eyes...they bleed! LGTM Some things that jumped out at me that may affect us:

1.  Changes to how ListWatchers are created
2.  Storage config in master has moved from fields to a map
3.  Change to endpoints from a string to a defined type

I can no longer read any secrets code without picturing @pmorie standing behind me whispering "shhh!"

File: Godeps/_workspace/src/github.com/GoogleCloudPlatform/kubernetes/pkg/api/rest/create_test.go (hunk 1 above)
Comment (smarterclayton) [inline Godeps/_workspace/src/github.com/GoogleCloudPlatform/kubernetes/pkg/api/rest/create_test.go]: I dunno, ask upstream.

**Comment by smarterclayton** [comment]
The router is fixed in my last commit.  Also, you don't have to review the merge commits :). Just my fixes.

**Comment by smarterclayton** [comment]
[test]

**Comment by deads2k** [comment]
I'll look at adding a proper error for #1135  today or Monday.  I previously misunderstood and thought that self-linking was already tolerant upstream.

The number of required changes is much smaller than I anticipated.  Is there going to be a card for someone to take advantage of the generic registry stuff?

**Comment by smarterclayton** [comment]
@deads2k calling ResourceAccessReviews("") is disallowed now upstream (you can't invoke create on a namespace scoped resource with ""), which causes:

```
--- FAIL: TestResourceAccessReview (3.91s)
	authorization_test.go:188: expected forbidden error, but didn't get one
	authorization_test.go:194: unexpected error: request construction error: 'an empty namespace may not be set during creation'
	authorization_test.go:199: expected [system:cluster-admins], got []
	authorization_test.go:202: expected [system:kube-client system:openshift-client system:openshift-deployer], got []
```

My preference is to define GlobalResourceAccessReview() method that takes no namespace and does not set Namespace() on the Create method.  Do you have an alternative suggestion?

**Comment by deads2k** [comment]
> My preference is to define GlobalResourceAccessReview() method that takes no namespace and does not set Namespace() on the Create method. Do you have an alternative suggestion

I'd have both.  A GlobalResourceAccessReview for cluster-admins who have rights to access resources across all namespaces and a ResourceAccessReview for project admins who only have rights to a single namespace.  I had heard rumors that global objects were going to be eliminated...

@derekwaynecarr Will we still have globally objects in a month?

**Comment by liggitt** [comment]
@smarterclayton fyi, I need the secrets stuff in this for #1208

**Comment by smarterclayton** [comment]
Brenton is going to cut beta2 before this lands.  Talk to him and make sure the other stuff for beta2 is ready to land.

**Comment by smarterclayton** [comment]
[merge]

**Pull request merged**